src/evo/assetlocktx.h

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2023 The Dash Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_EVO_ASSETLOCKTX_H
+#define BITCOIN_EVO_ASSETLOCKTX_H
+
+#include <bls/bls_ies.h>
+#include <evo/specialtx.h>
+#include <primitives/transaction.h>
+
+#include <key_io.h>
+#include <serialize.h>
+#include <tinyformat.h>
+#include <univalue.h>
+
+class CBlockIndex;
+struct CCreditPool;
+
+class CAssetLockPayload
+{
+public:
+    static constexpr uint8_t CURRENT_VERSION = 1;
+    static constexpr auto SPECIALTX_TYPE = TRANSACTION_ASSET_LOCK;
+
+private:
+    uint8_t nVersion{CURRENT_VERSION};
+    std::vector<CTxOut> creditOutputs;
+
+public:
+    explicit CAssetLockPayload(const std::vector<CTxOut>& creditOutputs) :
+        creditOutputs(creditOutputs)
+    {}
+
+    CAssetLockPayload() = default;
+
+    SERIALIZE_METHODS(CAssetLockPayload, obj)
+    {
+        READWRITE(
+            obj.nVersion,
+            obj.creditOutputs
+        );
+    }
+
+    std::string ToString() const;
+
+    void ToJson(UniValue& obj) const
+    {
+        obj.clear();
+        obj.setObject();
+        obj.pushKV("version", int(nVersion));
+        UniValue outputs;
+        outputs.setArray();
+        for (const CTxOut& out : creditOutputs) {
+            outputs.push_back(out.ToString());
+        }
+        obj.pushKV("creditOutputs", outputs);
+    }
+
+    // getters
+    uint8_t getVersion() const;
+    const std::vector<CTxOut>& getCreditOutputs() const;
+};
+
+class CAssetUnlockPayload
+{
+public:
+    static constexpr uint8_t CURRENT_VERSION = 1;
+    static constexpr auto SPECIALTX_TYPE = TRANSACTION_ASSET_UNLOCK;
+
+    static constexpr size_t MAXIMUM_WITHDRAWALS = 32;
```

Comment:
accordingly DIP:

> In order to limit the signing load on validators, there can only be a maximum of 16 withdrawal transactions per platform block. However, with transaction pooling at a maximum of 32 withdrawals per transaction, this can amount to up to 512 withdrawals per block. Assuming a 5 second block this would lead to about 102 withdrawals per second.